source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md

```diff
@@ -0,0 +1,194 @@
+  assert.equal(userCanVote, true);
+});</code></pre>
+
+Wonderful, we have 5 passing tests now!
+
+## Testing `vote()`
+
+Of course we want to test whether one of our application's core features works as well. There's certainly different ways to verify whether `vote()` does what it's supposed to do, but for this tutorial we'll simply check whether the owner account of the vote emitted by the `NewVote` event is the same as the account that performed the vote. We can actually take some inspiration from our previous tests:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+  ...
+  it("should be able to vote in a post", async () => {
+    const receipt = await DReddit.methods.vote(postId, 1).send();
+    const Vote = receipt.events.NewVote;
+    assert.equal(Vote.returnValues.owner, accounts[0]);
+  });
+});</code></pre>
+
+## Test that only one vote per post is allowed
+
+The last but essential functionality we want to test is that whether our Smart Contract allows users to vote multiple times on the same post, which for obvious reasons shouldn't be possible. Using the `async/await` syntax we can test this very nicely by adding a `try/catch` block. When a user votes on a post she has already voted on, `vote()` will fail in which case we can make our assertions accordingly:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+  ...
+  it('shouldn\'t be able to vote twice', async () => {
+    try {
+      const receipt = await DReddit.methods.vote(postId, 1).send();
+      assert.fail('should have reverted');
+    } catch (error){
+      assert(error.message.search('revert') > -1, 'Revert should happen');
+    }
+  });
+});</code></pre>
+
+This might look a bit confusing first but it's actually pretty straight forward. In case `vote()` fails, we should not reach the `assert.fail()` call but end up in the `catch()` block immediately. If that was not the case, the test would fail. This is a very common pattern when testing negatives.
+
+Okay, one last time we run `embark test` and if the output looks like the following, we're fully covered in terms of tests!
+
+
+<pre class="shell"><code>❯ embark test
+Compiling contracts
+
+
+  DReddit
+    ✓ should work (0ms) - [0 gas]
+    ✓ should be able to create a post and receive it via contract event (60ms) - [160689 gas]
+    ✓ post should have correct data (18ms) - [0 gas]
+    ✓ should not be able to vote in an unexisting post (14ms) - [0 gas]
+    ✓ should be able to vote in a post if account hasn't voted before (12ms) - [0 gas]
+    ✓ should be able to vote in a post (42ms) - [65115 gas]
+    ✓ shouldn't be able to vote twice (37ms) - [22815 gas]
+
+
+  7 passing (5s) - [Total: 3130955 gas]
+
+ > All tests passed</code></pre>
+ 
+ Awesome! If you run into any issues, check out the repository with all steps recorded [here](TODO). In the next and last part of this series, we'll be building a front-end for our DReddit app using React. Until then, feel free to add more tests as you like!
```

Comment:
Nice. Was just gonna recommend putting a link to the repo where all the tests are implemented.

Reply:
Done!